### ISSUE TITLE
Should devicemapper context be obtained every time it's needed?

### ISSUE THREAD

**Issue by mulkieran** [stratis-storage/stratisd#265]
Or should it be stashed the first time. I imagine that getting the context is expensive, but I'm not sure.

**Comment by agrover** [comment]
It's making a system call to open a file. Not a huge deal but still, best not to be flagrant.

Can we reuse it within each top-level Engine API call, perhaps?

@trgill I think you were also considering what to do here, if you have any thoughts.

**Comment by trgill** [comment]
Yeah, it looks like your suggestion to reuse with each top-level engine API is the most idiomatic approach.  Looks like singletons are frowned upon in Rust.

**Comment by agrover** [comment]
some places pass around the DM -- we need to do a sweep to be consistent about this.

**Comment by mulkieran** [comment]
If the file is open, are others prevented from using devicemapper? It would be nice just to initialize the context when the engine starts up and then get it whenver it is needed, but that would be a bad idea if we're preventing others from accessing it. @agrover @trgill?

**Comment by agrover** [comment]
No they are not prevented, it can be opened multiple times.

**Comment by mulkieran** [comment]
I considered using ```sync::Once``` but failed due to being too close to the cutting edge.

We want to:
1. Handle an error return intelligently at startup.
2. Not have to unwrap() all regular calls.

**Comment by agrover** [comment]
make `DM` a member of `StratEngine` and acquire it in `StratEngine::initialize`?

**Comment by mulkieran** [comment]
We are quite consistent with this. We acquire the DM context exactly in methods that implement public traits. Otherwise, we pass it as an argument. Any one of the public methods could be called directly by the API layer, which should not be responsible for handling DM context.

**Comment by agrover** [comment]
ok cool. So should this still be an open issue?

**Comment by mulkieran** [comment]
Closed by #858
